cmd/authority-claimer/src/repository.rs

```diff
@@ -0,0 +1,205 @@
+
+/// The `Repository` queries the database and gets an unsubmitted claim
+#[async_trait]
+pub trait Repository: Debug {
+    type Error: snafu::Error + 'static;
+
+    async fn get_claim(
+        &mut self,
+    ) -> Result<(RollupsClaim, Address), Self::Error>;
+
+    async fn update_claim(
+        &mut self,
+        id: u64,
+        tx_hash: H256,
+    ) -> Result<(), Self::Error>;
+}
+
+#[derive(Debug, Snafu)]
+#[snafu(visibility(pub(super)))]
+pub enum RepositoryError {
+    #[snafu(display("database error"))]
+    DatabaseSqlx { source: sqlx::Error },
+
+    #[snafu(display("int conversion error"))]
+    IntConversion { source: std::num::TryFromIntError },
+}
+
+#[derive(Clone, Debug)]
+pub struct DefaultRepository {
+    // Connection is not thread safe to share between threads, we use connection pool
+    db_pool: Arc<Pool<Postgres>>,
+}
+
+impl DefaultRepository {
+    /// Create database connection pool, wait until database server is available
+    pub fn new(endpoint: String) -> Result<Self, RepositoryError> {
+        let connection = PgPoolOptions::new()
+            .acquire_timeout(REPOSITORY_ACQUIRE_TIMEOUT)
+            .min_connections(REPOSITORY_MIN_CONNECTIONS)
+            .max_connections(REPOSITORY_MAX_CONNECTIONS)
+            .connect_lazy(&endpoint)
+            .context(DatabaseSqlxSnafu)?;
+        Ok(Self {
+            db_pool: Arc::new(connection),
+        })
+    }
+
+    /// Obtain the connection from the connection pool
+    async fn conn(&self) -> PoolConnection<Postgres> {
+        self.db_pool
+            .acquire()
+            .await
+            .expect("No connections available in the pool")
+    }
+}
+
+#[async_trait]
+impl Repository for DefaultRepository {
+    type Error = RepositoryError;
+
+    async fn get_claim(
+        &mut self,
+    ) -> Result<(RollupsClaim, Address), Self::Error> {
+        let claim: RollupsClaim;
+        let iconsensus_address: Address;
+        let mut conn = self.conn().await;
+        let (sql, values) = Query::select()
+            .columns([
+                (Epoch::Table, Epoch::Id),
+                (Epoch::Table, Epoch::ClaimHash),
+                (Epoch::Table, Epoch::ApplicationAddress),
+                (Epoch::Table, Epoch::LastBlock),
+            ])
+            .column((Application::Table, Application::IconsensusAddress))
+            .from(Epoch::Table)
+            .inner_join(
+                Application::Table,
+                Expr::col((Epoch::Table, Epoch::ApplicationAddress))
+                    .equals((Application::Table, Application::ContractAddress)),
+            )
+            .and_where(Expr::col((Epoch::Table, Epoch::Status)).eq(
+                Func::cast_as("CLAIM_COMPUTED", Alias::new("\"EpochStatus\"")),
+            ))
+            .order_by(Epoch::Index, Order::Asc)
+            .order_by((Epoch::Table, Epoch::Id), Order::Asc)
+            .limit(1)
+            .build_sqlx(PostgresQueryBuilder);
+
+        loop {
```

Comment:
We should revisit this strategy for the final Golang claimer.

Reply:
I've already discussed with @renan061 another strategy for the final golang claimer, which solves this issue better